Add configuration for taxes

razvanudream opened the pull request:
**What problem is this solving?**

This PR is meant to add to the prices sent to google the taxes (that's for stores when the prices are initially without a tax and the tax is separate in commertialOffer (the fields we're interested in are Price, spotPrice and Tax)

**How should this be manually tested?**

I created a configuration in the app's settings: https://structured--dacris.myvtex.com/admin/apps/vtex.structured-data@0.6.2/setup/

By default the decimals value is set to 2 and pricesWithTax is set to false to not impact any store that's currently using this app. Keep in mind that even though in manifest.json I set a default value, when initially installing the app, if you don't go to the app's settings and press Save changes (even without changing the default values) they will return as null from graphQL, as a fix for that scenario I set them in constants within Product.js in case they return null

OK -> install Chrome's extension for Structured Data and then go to a product

https://structured--dacris.myvtex.com/masca-faciala-de-uz-medical-tip-i-50buc-cut/p?__bindingAddress=www.dacris.net/

Generate the product response using the plugin and the lowPrice/highPrice/price should be either with the tax added or without, depending on which settings you configured.

As for the decimals, if you set it to 3 for example and the number is 6.239999 it won't say 6.240 it will say 6.24 (just keep that in mind when testing), it does work however for numbers such as 6.23333 (output will be 6.233)

**Screenshots or example usage:**

For this product

![image](https://user-images.githubusercontent.com/71461884/124725653-b0e46f00-df15-11eb-82fe-492fcf192383.png)

With tax added (in this case it's 19% but the Tax value from commertialOffer is a float number so it's added to the price)

![image](https://user-images.githubusercontent.com/71461884/124725349-75e23b80-df15-11eb-8dc1-6c000caedb42.png)

Without tax:

![image](https://user-images.githubusercontent.com/71461884/124725600-a629da00-df15-11eb-888f-309429726e4b.png)

lariciamota commented on react/Product.js:
```diff
@@ -93,8 +116,8 @@ const composeAggregateOffer = (product, currency) => {
     '@type': 'AggregateOffer',
-    lowPrice: getSpotPrice(low),
-    highPrice: getPrice(high),
+    lowPrice: pricesWithTax ? Math.round(((getSpotPrice(low) + getTax(low)) + Number.EPSILON) * Math.pow(10, decimals)) / Math.pow(10, decimals) : getSpotPrice(low),
+    highPrice: pricesWithTax ? Math.round(((getPrice(high) + getTax(high)) + Number.EPSILON) * Math.pow(10, decimals)) / Math.pow(10, decimals) : getPrice(high),
```
You could have a function like
```
const getFinalPrice = (value, decimals, getPriceFunc) => {
  return pricesWithTax ? Math.round(((getPriceFunc(value) + getTax(value)) + Number.EPSILON) * Math.pow(10, decimals)) / Math.pow(10, decimals) : getPriceFunc(value)
}
```

And then it'd be simplified to:
```suggestion
    lowPrice: getFinalPrice(low, decimals, getSpotPrice),
    highPrice: getFinalPrice(high, decimals, getPrice),
```
What do you think about it?

razvanudream replied:
That's a nice change to make the code readable as well as avoid duplicating it.  :) Made this change as well.  As for the get settings query, I thought I could do it with https://github.com/vtex-apps/render-runtime#getsettings but when I used vtex.structured-data instead of vtex.store it returned null. Either I'm reading the docs wrong or it just works with the vtex.store app which returns store related settings that are of no use to me in this case.  Let me know if you know of a better way to get the app's settings though, thanks.

lariciamota commented on react/Product.js:
```diff
@@ -35,15 +56,18 @@ const lowHighForSellers = (sellers) => {
   const availability = getSKUAvailabilityString(seller)
-  const price = getSpotPrice(seller)
+
+  const price = pricesWithTax ? Math.round(((getSpotPrice(seller) + getTax(seller)) + Number.EPSILON) * Math.pow(10, decimals)) / Math.pow(10, decimals) : getSpotPrice(seller)
```
It could also use the function I suggested

lariciamota reviewed (changes requested):
Thank you for your contribution!

I'm under the impression that the settings handling can be simplified, so I'll ask for the help of someone with more context about it 😄 
In the meantime, please check the suggestions I commented.

razvanudream commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+- Added `settingsSchema` to allow stores to send `pricesWithTax` and specify the number of `decimals`
```
Updated the changelog

razvanudream commented:
I noticed the failing tests and lint errors, most of them are related to the useQuery implementation to get the app's settings. Since there might be a better way to get that info, I will fix the tests once we gave a definitive answer

lariciamota commented:
> I noticed the failing tests and lint errors, most of them are related to the useQuery implementation to get the app's settings. Since there might be a better way to get that info, I will fix the tests once we gave a definitive answer

Hey @razvanudream !
You did it correctly in this case, we can proceed using the GraphQL query 😄

razvanudream commented:
@emersonlaurentino - thanks for the review! I fixed the lint errors and the failing tests and also a few other things:

1. the sortByPriceAsc function was split in 2, one for the spotPrice without tax and one with the tax added, IMHO it makes the code easier to follow and also keeps things separate in terms of what a function does and returns

2. the decimal and pricesWithTax returned from the app's settings, in case of undefined the default will be set directly in the getAppSettings function, this way I don't have to use the short if syntax and it's also easier to understand

3. the price on line 73 (now) should use getSpotPrice instead of getPrice (accidentally changed that one during the first commit)

4. Removed locale from StructuredData because it was unused :)

igorbrasileiro commented on react/Product.js:
```diff
@@ -5,19 +5,45 @@ import PropTypes from 'prop-types'
 // eslint-disable-next-line no-restricted-imports
 import { pathOr, path, sort, last, flatten } from 'ramda'
 import { jsonLdScriptProps } from 'react-schemaorg'
+import { useQuery } from 'react-apollo'
```
lint here

igorbrasileiro commented on react/Product.js:
```diff
@@ -5,19 +5,45 @@ import PropTypes from 'prop-types'
+const DEFAULT_PRICES_WITH_TAX = false
+
+const getAppSettings = () => {
+  const { data } = useQuery(GET_SETTINGS, { ssr: false })
```
Hooks can't be used outside the react tree and the order that hooks are called, matters.

razvanudream replied:
Capitalized it to be able to use it as described in https://reactjs.org/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized, I see no other way around it. I need it there to be able to also mock it and test it (not the query itself, but the function that uses its response.

victorhmp replied:
Sorry, but this seems like cheating to me 😅 . I really don't think this is the best approach. 

If the only reason you cannot call this hook inside of the `StructuredData` component is because of testing tools, then there's something wrong with the testing tools. 

Buuut, in general, using `@vtex/test-tools`, you should be able to mock and test GraphQL queries without any issues. We do it in this app for example: https://github.com/vtex-apps/product-gifts/blob/master/react/__tests__/ProductGifts.test.tsx. Check out how we create this `renderWithProductContext` utility function: https://github.com/vtex-apps/product-gifts/blob/22fb0e4ad55604c609b3621b1d4e29c9135a247a/react/__mocks__/testUtils.tsx#L17. Something similar to this should work here :)

razvanudream replied:
> Sorry, but this seems like cheating to me  . I really don't think this is the best approach.
> 
> If the only reason you cannot call this hook inside of the `StructuredData` component is because of testing tools, then there's something wrong with the testing tools.
> 
> Buuut, in general, using `@vtex/test-tools`, you should be able to mock and test GraphQL queries without any issues. We do it in this app for example: https://github.com/vtex-apps/product-gifts/blob/master/react/__tests__/ProductGifts.test.tsx. Check out how we create this `renderWithProductContext` utility function: https://github.com/vtex-apps/product-gifts/blob/22fb0e4ad55604c609b3621b1d4e29c9135a247a/react/__mocks__/testUtils.tsx#L17. Something similar to this should work here :)

Hi @victorhmp - I see what you're saying, unfortunately I'm not very familiar with tests in React, only learned it a year ago and before that I only had experience with PHP/Python. I can't figure out a quick solution on how to mock graphQL into the test, so if you're not happy with the "workaround" that passes the test and has no impact on code execution whatsoever (AFAIK), could you please do the changes yourself?  I changed the manifest to accept i18n messages and would help a lot if you could offer some assistance for the test so we could merge this ASAP.   The client's been waiting for a week now...

Thanks

icazevedo replied:
This is really an issue. You have to transform it to a hook (`useAppSettings`) and provide a default implementation as a mock for the testing suite. You can make it return a fixed value or have multiple implementations for it depending on if you want to test the tax and decimals behavior.

igorbrasileiro commented on react/Product.js:
```diff
@@ -5,19 +5,45 @@ import PropTypes from 'prop-types'
 
-const lowHighForSellers = (sellers) => {
-  const sortedByPrice = sortByPriceAsc(sellers)
+const lowHighForSellers = (sellers, pricesWithTax) => {
```
Add the second parameter as object, it will improve maintainability. Also type this object

```suggestion
interface LowHighForSellersOptions {
  ...
}
const lowHighForSellers = (sellers, { pricesWithTax }: LowHighForSellersOptions) => {
```

razvanudream replied:
Done, added the object but since it's a JS file, interfaces are not allowed. Only allowed in tsx files AFAIK

igorbrasileiro replied:
Yeah, my bad, I didn't see that's a js file.

igorbrasileiro reviewed (changes requested):
Great job @razvanudrem, left some comments.
Isn't good to have functions with many params, so putting all params inside an object can have multiple benefits. Can you do this? Also, lint all files you changed.

igorbrasileiro commented:
Also if you can type new parameters that you added, would be good.

victorhmp commented on manifest.json:
```diff
@@ -14,7 +14,26 @@
+    "type": "object",
+    "properties": {
+      "decimals": {
+        "title": "Number of decimals",
+        "type": "number",
+        "default": 2,
+        "description": "Set the number of decimals you want the prices to show"
+      },
+      "pricesWithTax": {
+        "title": "Prices with tax",
+        "type": "boolean",
+        "default": false,
+        "description": "Use the Price with Tax instead of spotPrice"
+      }
```
You should use message ids here to make titles and descriptions i18n supported

razvanudream replied:
Sure, but don't know how to do that in manifest. I know how to in a tsx file using react-intl, but can't find anything on using the message builder and the message ids in a settingsSchema file (https://developers.vtex.com/vtex-developer-docs/docs/vtex-io-documentation-4-configuringyourappsettings) - can you show an example of how to set the descriptions and titles in this case to allow multiple languages?

victorhmp replied:
Yep, check how we did this in our `vtex.store` app -> https://github.com/vtex-apps/store/blob/master/manifest.json#L39.

You just reference the message IDs in the same way you would inside a component

razvanudream replied:
@victorhmp - I noticed that when I add the message ids in manifest they don't display properly in admin, I tried in other apps of mine with admin builder and the same happens.  What's different between this app and store app that seems to be using the messages from admin-pages

![image](https://user-images.githubusercontent.com/71461884/127622485-a132906a-4a13-4471-9c79-12a3814570ed.png)

victorhmp reviewed (changes requested):
Hi @razvanudream! Please make sure to lint these files

razvanudream commented:
Hi @icazevedo & @victorhmp - I added the hook for useAppSettings and also created a mock for it in the Product test as well as a test for pricesWithTax set to true instead of the default false. Let me know what you think (I used jest)

igorbrasileiro commented on react/__tests__/Product.test.js:
```diff
@@ -4,6 +4,16 @@ import { parseToJsonLD } from '../Product'
+let mockDecimals = 2
+let mockPricesWithTax = false
+
+jest.mock('../hooks/useAppSettings', () => {
+  return jest.fn(() => ({
+    decimals: mockDecimals,
+    pricesWithTax: mockPricesWithTax,
+  }))
+})
+
```
It seems not to be used. Please remove it

igorbrasileiro reviewed (changes requested):
Remove unused code in the test file. Also, there is no test for the new hook `useAppSettings`

razvanudream commented:
@igorbrasileiro - Ok, I created the useAppSettings.test.jsx that basically tests that the JSON.parse works on the mocked data object that mimics the request and result of the useQuery, since I return an object containing decimals and pricesWithTax (without loading or error), I can't test those as well, expect them to be undefined or false, etc. but I think this is good enough. As for the Product.test.js, removed the jest part and left the extra test there to cover case when pricesWithTax is set to true

icazevedo commented on react/hooks/useAppSettings.ts:
```diff
@@ -0,0 +1,33 @@
+
+  const { data, loading } = useQuery(GET_SETTINGS, { ssr: false })
+
+  useEffect(() => {
+    if (!loading) {
+      const { decimals, pricesWithTax } = JSON.parse(data?.appSettings?.message)
+
+      setSettings({ decimals, pricesWithTax })
+    }
+  }, [loading, data])
```
Oh, I'm sorry for the early approval. I think this can break if the request fails or if there's nothing on the settings. I think it's better to check it before deconstructing.

```suggestion
  useEffect(() => {
    if (!loading && data?.appSettings?.message) {
      const { decimals, pricesWithTax } = JSON.parse(data.appSettings.message)
      
      if(decimals !== undefined && decimals !== null && pricesWithTax !== null && pricesWithTax !== undefined) {
        setSettings({ decimals, pricesWithTax })
      }
    }
  }, [loading, data, setSettings])
```

razvanudream replied:
You're right - deconstructing works regardless of decimals and pricesWithTax being defined, but I added a check after to setSettings state only if they are defined, otherwise the state will stay as it was initially set. This problem appears because the default value in manifest.json for settingsSchema fields only works if you go to the app's settings and press the save changes button once (without changing anything), that's when the default values you have defined in manfiest are applied. Until then they're useless.  Hence why I had to set the initial state to those default values.

Good catch, thanks!

icazevedo commented on react/hooks/useAppSettings.ts:
```diff
@@ -0,0 +1,33 @@
+    if (!loading) {
+      const { decimals, pricesWithTax } = JSON.parse(data?.appSettings?.message)
+
+      if (decimals && pricesWithTax) setSettings({ decimals, pricesWithTax })
```
This will fail if `pricesWithTax` is `false`, which is not the desired behavior. Also, if `decimals` is `0`, it will also fail.

icazevedo commented on react/hooks/useAppSettings.ts:
```diff
@@ -0,0 +1,33 @@
+
+  useEffect(() => {
+    if (!loading) {
+      const { decimals, pricesWithTax } = JSON.parse(data?.appSettings?.message)
```
Try to run `JSON.parse(undefined)` or `JSON.parse('')` on your browser. It will throw an error. You really have to check it before to make sure it won't throw a error due to foreseeable causes.

razvanudream replied:
Fair enough - Added better handling and an extra test for null values for piece of mind :) Cheers

icazevedo replied:
Again, I think you're getting the wrong picture here: if `data`, `appSettings`, or `message` are `null` or `undefined` or `''`, this will throw an error. You can't parse those values and even if you could, you couldn't deconstruct them.

razvanudream replied:
Sorry, long week... I see what you're saying, it should be OK now
